mllib/src/test/scala/org/apache/spark/mllib/evaluation/MulticlassMetricsSuite.scala

```diff
@@ -95,4 +95,95 @@ class MulticlassMetricsSuite extends SparkFunSuite with MLlibTestSparkContext {
+
+    assert(metrics.confusionMatrix.toArray.sameElements(confusionMatrix.toArray))
+    assert(math.abs(metrics.truePositiveRate(0.0) - tpRate0) < delta)
+    assert(math.abs(metrics.truePositiveRate(1.0) - tpRate1) < delta)
+    assert(math.abs(metrics.truePositiveRate(2.0) - tpRate2) < delta)
+    assert(math.abs(metrics.falsePositiveRate(0.0) - fpRate0) < delta)
+    assert(math.abs(metrics.falsePositiveRate(1.0) - fpRate1) < delta)
+    assert(math.abs(metrics.falsePositiveRate(2.0) - fpRate2) < delta)
+    assert(math.abs(metrics.precision(0.0) - precision0) < delta)
+    assert(math.abs(metrics.precision(1.0) - precision1) < delta)
+    assert(math.abs(metrics.precision(2.0) - precision2) < delta)
+    assert(math.abs(metrics.recall(0.0) - recall0) < delta)
+    assert(math.abs(metrics.recall(1.0) - recall1) < delta)
+    assert(math.abs(metrics.recall(2.0) - recall2) < delta)
+    assert(math.abs(metrics.fMeasure(0.0) - f1measure0) < delta)
+    assert(math.abs(metrics.fMeasure(1.0) - f1measure1) < delta)
+    assert(math.abs(metrics.fMeasure(2.0) - f1measure2) < delta)
+    assert(math.abs(metrics.fMeasure(0.0, 2.0) - f2measure0) < delta)
+    assert(math.abs(metrics.fMeasure(1.0, 2.0) - f2measure1) < delta)
+    assert(math.abs(metrics.fMeasure(2.0, 2.0) - f2measure2) < delta)
+
+    assert(math.abs(metrics.accuracy -
+      (2.0 * w1 + 2.0 * w1 + 1.0 * w2 + 1.0 * w2) / tw) < delta)
+    assert(math.abs(metrics.accuracy - metrics.precision) < delta)
+    assert(math.abs(metrics.accuracy - metrics.recall) < delta)
+    assert(math.abs(metrics.accuracy - metrics.fMeasure) < delta)
+    assert(math.abs(metrics.accuracy - metrics.weightedRecall) < delta)
+    assert(math.abs(metrics.weightedTruePositiveRate -
+      (((2 * w1 + 1 * w2 + 1 * w1) / tw) * tpRate0 +
+        ((1 * w2 + 2 * w1 + 1 * w2) / tw) * tpRate1 +
+        (1 * w2 / tw) * tpRate2)) < delta)
+    assert(math.abs(metrics.weightedFalsePositiveRate -
+      (((2 * w1 + 1 * w2 + 1 * w1) / tw) * fpRate0 +
+        ((1 * w2 + 2 * w1 + 1 * w2) / tw) * fpRate1 +
+        (1 * w2 / tw) * fpRate2)) < delta)
+    assert(math.abs(metrics.weightedPrecision -
+      (((2 * w1 + 1 * w2 + 1 * w1) / tw) * precision0 +
+        ((1 * w2 + 2 * w1 + 1 * w2) / tw) * precision1 +
+        (1 * w2 / tw) * precision2)) < delta)
+    assert(math.abs(metrics.weightedRecall -
+      (((2 * w1 + 1 * w2 + 1 * w1) / tw) * recall0 +
+        ((1 * w2 + 2 * w1 + 1 * w2) / tw) * recall1 +
+        (1 * w2 / tw) * recall2)) < delta)
+    assert(math.abs(metrics.weightedFMeasure -
+      (((2 * w1 + 1 * w2 + 1 * w1) / tw) * f1measure0 +
+        ((1 * w2 + 2 * w1 + 1 * w2) / tw) * f1measure1 +
+        (1 * w2 / tw) * f1measure2)) < delta)
+    assert(math.abs(metrics.weightedFMeasure(2.0) -
+      (((2 * w1 + 1 * w2 + 1 * w1) / tw) * f2measure0 +
+        ((1 * w2 + 2 * w1 + 1 * w2) / tw) * f2measure1 +
+        (1 * w2 / tw) * f2measure2)) < delta)
```

Comment:
There're many repeated expressions here such as `(2 * w1 + 1 * w2 + 1 * w1) / tw`, could you store them in variables first ?